examples/ota-requestor-app/linux/main.cpp

```diff
@@ -43,44 +46,107 @@
                           chip::ByteSpan updateToken, bool userConsentNeeded, chip::ByteSpan metadataForRequestor)
 {
     ChipLogDetail(SoftwareUpdate, "%s", __FUNCTION__);
+
+    TransferSession::TransferInitData initOptions;
+    initOptions.TransferCtlFlags = chip::bdx::TransferControlFlags::kReceiverDrive;
+    initOptions.MaxBlockSize     = 1024;
+    char testFileDes[9]          = { "test.txt" };
+    initOptions.FileDesLength    = static_cast<uint16_t>(strlen(testFileDes));
+    initOptions.FileDesignator   = reinterpret_cast<uint8_t *>(testFileDes);
+
+    {
+        chip::Messaging::ExchangeManager * exchangeMgr = providerDevice->GetExchangeManager();
+        chip::Optional<chip::SessionHandle> session    = providerDevice->GetSecureSession();
+        if (exchangeMgr != nullptr && session.HasValue())
+        {
+            exchangeCtx = exchangeMgr->NewContext(session.Value(), &bdxDownloader);
+        }
+
+        if (exchangeCtx == nullptr)
+        {
+            ChipLogError(BDX, "unable to allocate ec: exchangeMgr=%p sessionExists? %u", exchangeMgr, session.HasValue());
+            return;
+        }
+    }
+
+    bdxDownloader.SetInitialExchange(exchangeCtx);
+
+    // This will kick of a timer which will regularly check for updates to the bdx::TransferSession state machine.
+    bdxDownloader.InitiateTransfer(&chip::DeviceLayer::SystemLayer(), chip::bdx::TransferRole::kReceiver, initOptions, 20000);
 }
 
-void OnFailure(void * context, uint8_t status)
+void OnQueryFailure(void * context, uint8_t status)
 {
-    ChipLogDetail(SoftwareUpdate, "Received failure response %d\n", (int) status);
+    ChipLogDetail(SoftwareUpdate, "QueryImage failure response %" PRIu8, status);
 }
 
 Callback<OtaSoftwareUpdateProviderClusterQueryImageResponseCallback> mQueryImageResponseCallback(OnQueryImageResponse, nullptr);
-Callback<DefaultFailureCallback> mOnFailureCallback(OnFailure, nullptr);
+Callback<DefaultFailureCallback> mOnQueryFailureCallback(OnQueryFailure, nullptr);
+void OnConnection(void * context, Device * device)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::Controller::OtaSoftwareUpdateProviderCluster cluster;
+    constexpr EndpointId kOtaProviderEndpoint = 0;
+
+    chip::Callback::Cancelable * successCallback = mQueryImageResponseCallback.Cancel();
+    chip::Callback::Cancelable * failureCallback = mOnQueryFailureCallback.Cancel();
+
+    // QueryImage params
+    constexpr VendorId testVendorId          = VendorId::Common;
+    constexpr uint16_t testProductId         = 77;
+    constexpr uint16_t testImageType         = 0;
+    constexpr uint16_t testHWVersion         = 3;
+    constexpr uint16_t testCurrentVersion    = 101;
+    constexpr uint8_t testProtocolsSupported = 0; // TODO: blocked because arrays are being generated as uint8_t
+    uint8_t locationBuf[3]                   = { 'U', 'S', '\0' };
```

Comment:
Please document why a null terminator is required here in a `uint8_t[]`

Reply:
This should be changed. A null-terminator isn't actually required for string arguments in cluster commands

However, the uint8_t is somewhat necessary because cluster commands accept string arguments as `ByteSpan` references